### ISSUE TITLE
head-enricher seems that is not worknig when using kafka as a transport layer

### ISSUE THREAD

**Issue by garciapd** [spring-projects/spring-xd-modules#24]
Following the same examples in the web, it is working if the transport is local but not if it is kafka.

Using the mode of kafka: embeddedHeaders
Using Spring XD 1.2.1
Using Kafka 2.10-0.8.2.1

**Comment by mbogoevici** [comment]
Thanks for reporting @garciapd

I've added this for tracking: https://jira.spring.io/browse/XD-3621

Please note that you would have to configure the headers you would like to propagate as part of application.yml (this feature is currently available for Redis but not for Kafka).

**Comment by garciapd** [comment]
Thanks for the quick response, I will follow the issue, it is ok as long as I can propagate it.

**Comment by garyrussell** [comment]
Just for completeness - the rabbit bus passes all headers by default (but they can be filtered if desired); the local bus passes all headers; the redis bus only passes certain headers by default (but the bus can be configured to pass specific headers). Kafka will be like redis when XD-3621 is resolved.

**Comment by garciapd** [comment]
For curiosity, why is the kafka going to behave as redis and not as rabbit?

**Comment by mbogoevici** [comment]
@garciapd 

Rabbit supports headers natively, Redis and Kafka don't - headers are embedded in the message. So in the latter case our approach is to propagate headers based on a configured whitelist. The issue here is that there's a bug that doesn't allow these headers to be configured for Kafka. 

Hope this helps.

**Comment by garciapd** [comment]
How can I get the standard headers? the correlationId is passed, and if I put the trackHistory=true I also get it, but I'm not able to get the id and the timestamp.

Thanks!

**Comment by garciapd** [comment]
Oh I see, the class KafkaMessageBus is only copying the STANDARD_HEADERS for org.springframework.xd.dirt.integration.bus.XdHeaders, shouldn't it be copying the MessageHeaders also defined in org.springframework.messaging.MessageHeaders?

**Comment by garyrussell** [comment]
The `id` and `timestamp` headers are unique to a message instance (they change each time a message is "changed" - but messages are never changed; they are immutable, a new message is created from the old one each time a message needs to be mutated).

So, even if we transferred them on the bus, the final message emitted on the other side will have different values in those headers.

Hence there is no point in transferring them.

If you wish to have some identifier and/or timestamp that is carried along across all message mutations you need to use some custom header(s).

**Comment by garciapd** [comment]
Yes understood and agreed, I'm looking at the history in order to get the timestamp (using the module.[modulename].trackHistory) and I realized that there is any.

What about the change you proposed? https://github.com/garyrussell/spring-xd/commit/ba15a1390f7e448dbc723ee76a45c2e239e0994e

I think it is very valuable to have the timestamp in the history, it tell you when it happened.

**Comment by garyrussell** [comment]
Wow - how did you find that? I have no recollection why that never made it into master; it certainly should have IMHO.

**Comment by garciapd** [comment]
Hehe yes I also think that the timestamp have to be included, should I create another issue with that? 

I'm trying to measure the latency between different processors in the stream in real time, using the history it is not a good option for production because it is adding an overhead but for debugging it is useful, but by having the timestamp :).  This is how I have ended up by looking at the implementation of the AbstractMessageBusBinderPlugin and realizing about your modification in the code.

**Comment by garyrussell** [comment]
https://jira.spring.io/browse/XD-3693

**Comment by garciapd** [comment]
Lovely! thanks

**Comment by garyrussell** [comment]
Bear in mind that when crossing server boundaries the timestamps might be misleading. NTP is your friend.

**Comment by garciapd** [comment]
Sure thanks!
